Cesium3DTilesSelection/src/upsampleGltfForRasterOverlays.cpp

```diff
@@ -1161,4 +1171,136 @@ static bool upsamplePrimitiveForRasterOverlays(
   return false;
 }
 
+namespace {
+struct BufferRangeToCopy {
+  int32_t sourceBuffer = -1;
+  int32_t targetBuffer = -1;
+  int64_t byteOffset = 0;
+  int64_t byteLength = 0;
+};
+} // namespace
+
+// Copy a buffer view from a parent to a child while accumulating a list of
+// unique buffers that are referenced. For each referenced buffer, track the
+// the range that needs to be copied. This lets us eliminate duplicate buffers
+// as well as reduce memcpy-ing sections of the buffer that aren't needed
+// by the buffer views (e.g. textures, attributes that will be reconstructed
+// anyways, etc).
+static int32_t createCopiedBufferView(
+    const Model& parentModel,
+    int32_t parentBufferViewId,
+    Model& result,
+    std::vector<BufferRangeToCopy>& bufferRanges) {
+
+  // Check invalid buffer view.
+  if (parentBufferViewId < 0 || static_cast<size_t>(parentBufferViewId) >=
+                                    parentModel.bufferViews.size()) {
+    return -1;
+  }
+
+  const BufferView& parentBufferView =
+      parentModel.bufferViews[static_cast<size_t>(parentBufferViewId)];
+
+  // Check invalid buffer.
+  if (parentBufferView.buffer < 0 ||
+      static_cast<size_t>(parentBufferView.buffer) >=
+          parentModel.buffers.size()) {
+    // Should we return a valid buffer view with an invalid buffer instead?
+    return -1;
+  }
+
+  for (BufferRangeToCopy& existingRange : bufferRanges) {
+    if (existingRange.sourceBuffer == parentBufferView.buffer) {
+      if (parentBufferView.byteOffset < existingRange.byteOffset) {
+        existingRange.byteOffset = parentBufferView.byteOffset;
+      }
+
+      if (parentBufferView.byteLength > existingRange.byteLength) {
+        existingRange.byteLength = parentBufferView.byteLength;
+      }
```

Comment:
I think this logic is not quite right. Let's say the existingRange has a byteOffset of 10 and a byteLength of 10. Then the parentBufferView has a byteOffset of 0 and also a byteLength of 10. You'd want to copy bytes 0-19, but instead would only copy 0-9.

I think it might be easier and not really any less efficient to simply copy the bufferViews, and allocate a new buffer for each, rather than try to keep the buffers intact. In pathological cases, multiple bufferViews could refer to the same bytes in a buffer, but that would be pathological and we don't need to put too much effort into making it as efficient as it can be.

Reply:
Ah yeah that part isn't correct. But that's a good point in general, maybe we can simplify it and just copy a different buffer for each metadata property buffer view (the vertex attributes will be reconstructed in the later step). The only case I was considering was interleaved property arrays. Each property array within a feature table will have the same count, so I figured it would be possible to interleave them. But now that appears to me to be a pathological case and maybe even explicitly disallowed, since there's no accessor.